sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,655 @@
+information provided as compile-time constant properties through all uses of the
+pointer unless noted otherwise.
+
+
+=== Representation of `annotated_ptr`
+
+`annotated_ptr` is a class template, parameterized by the type of the underlying
+allocation `T`, and a list of properties `PropertyListT`.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+template <typename T, typename PropertyListT = properties<>>
+class annotated_ptr {
+  ...
+----
+
+`PropertyListT` enables properties to be associated with an `annotated_ptr`.
+Properties may be specified for an `annotated_ptr` to provide semantic
+modification or optimization hint information.
+
+Here's an example of how a property could be used:
+
+[source,c++]
+----
+using namespace sycl::ext::oneapi::experimental;
+{
+  sycl::queue q;
+  // A pointer to an integer
+  auto ptr = ...
+  // alignment of the pointer in bytes specified using the property 'alignment'
+  auto arg_a = annotated_ptr(ptr, properties{alignment<4>});
+
+  q.submit([=]{
+    ...
+    *arg_a = (*arg_a) * 2;
+  });
+  ...
+}
+----
+
+The section below and the table that follows, describe the constructors, member
+functions and factory methods for `annotated_ptr`.
+The section below refers to an `annotated_ref` class which is described in the
+section following this one.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+template <typename T, typename PropertyListT = properties<>>
+class annotated_ptr {
+  public:
+    using reference = annotated_ref<T, PropertyListT>;
+
+    annotated_ptr() noexcept;
+    explicit annotated_ptr(T *Ptr, const properties<PropertyListT> &) noexcept;
+    annotated_ptr(annotated_ptr const &) noexcept;
+    template <T, typename P> explicit annotated_ptr(
+      annotated_ptr<T, P> const &) noexcept;
+    template <T, typename PropertyListU, typename PropertyListV>
+    explicit annotated_ptr(annotated_ptr<T, PropertyListU> const &,
+      properties<PropertyListV>) noexcept;
+
+    reference operator*() const noexcept;
+    reference operator[](std::ptrdiff_t) const noexcept;
+    annotated_ptr operator+(size_t) const noexcept;
+    std::ptrdiff_t operator-(annotated_ptr) const noexcept;
+
+    operator bool() const noexcept;
+
+    // Implicit conversion is not supported
+    operator T*() noexcept = delete;
+    operator const T*() const noexcept = delete;
+
+    T* get() noexcept;
+    const T* get() const noexcept;
+
+    annotated_ptr& operator=(const T*) noexcept;
+    annotated_ptr& operator=(annotated_ptr const&) noexcept;
+
+    annotated_ptr& operator++() noexcept;
+    annotated_ptr operator++(int) noexcept;
+    annotated_ptr& operator--() noexcept;
+    annotated_ptr operator--(int) noexcept;
+
+    template<typename propertyT>
+    static constexpr bool has_property();
+
+    // The return type is an unspecified internal class used to represent
+    // instances of propertyT
+    template<typename propertyT>
+    static constexpr /*unspecified*/ get_property();
+
+  private:
```

Comment:
This should probably be left as an implementation detail even though it's the obvious way to implement it.